src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -2969,77 +2958,49 @@ private Keyspace getValidKeyspace(String keyspaceName)
      * Remove the snapshot with the given name from the given keyspaces.
      * If no tag is specified we will remove all snapshots.
      */
-    public void clearSnapshot(String tag, String... keyspaceNames)
+    public void clearSnapshot(String tag, String... keyspaceNames) throws IOException
     {
         clearSnapshot(Collections.emptyMap(), tag, keyspaceNames);
     }
 
     public void clearSnapshot(Map<String, Object> options, String tag, String... keyspaceNames)
     {
-        if (tag == null)
-            tag = "";
-
         if (options == null)
             options = Collections.emptyMap();
 
-        Set<String> keyspaces = new HashSet<>();
-        for (String dataDir : DatabaseDescriptor.getAllDataFileLocations())
-        {
-            for (String keyspaceDir : new File(dataDir).tryListNames())
-            {
-                // Only add a ks if it has been specified as a param, assuming params were actually provided.
-                if (keyspaceNames.length > 0 && !Arrays.asList(keyspaceNames).contains(keyspaceDir))
-                    continue;
-                keyspaces.add(keyspaceDir);
-            }
-        }
+        Set<String> keyspaces = new HashSet<>(Arrays.asList(keyspaceNames));
+        long maxCreatedAt = getMaxSnapshotCreatedAt(options);
+
+        SnapshotManager.instance.clearSnapshots(tag, keyspaces, maxCreatedAt);
 
+        if (logger.isDebugEnabled())
+            logger.debug("Cleared out snapshot directories");
```

Comment:
maybe an opportunity to improve this log message:
```suggestion
            logger.debug("Cleared out snapshot directories tag={} keyspaces={} maxCreatedAt={}", tag, keyspaceNames, maxCreatedAt);
```